File: idexal/ai-core/src/providers/openaiCompat.ts
// Idexal AI Core — OpenAI-compatible provider
// Works with any OpenAI-compatible endpoint: OpenAI, OpenRouter, Groq,
// DeepSeek, Mistral, Together, Ollama, LM Studio, vLLM, and any custom base URL.

import { httpJson, bearerToken, truncate } from './http.ts';
import {
	type ChatRequest,
	type ChatResponse,
	type EmbeddingsRequest,
	type LLMProvider,
	type Message,
	type ModelCapabilities,
	type StreamDelta,
	type TokenUsage,
	type ToolCall,
	type ToolDefinition,
	ProviderError,
} from './types.ts';

export interface OpenAiCompatOptions {
	id: string;
	displayName: string;
	baseUrl: string; // e.g. https://api.openai.com/v1 or http://localhost:11434/v1
	apiKey?: string;
	defaultModel: string;
	/** Extra headers merged into every request (custom provider support). */
	headers?: Record<string, string>;
	/** Environment variable that holds the API key (for custom providers). */
	apiKeyEnv?: string;
	capabilities?: Partial<ModelCapabilities>;
	/** Custom OpenAI-format request body fields merged in. */
	extraBody?: Record<string, unknown>;
	/** Optional function to transform a request body before sending. */
	transformRequest?: (body: Record<string, unknown>) => Record<string, unknown>;
	/** Dedicated embeddings model (e.g. text-embedding-3-small). Defaults to defaultModel. */
	embeddingModel?: string;
	/** Request timeout in ms (default 120s). Gateways may be slower. */
	timeoutMs?: number;
	/** True when the endpoint works without an API key (anonymous tier). */
	anonymous?: boolean;
}

interface OpenAiMessage {
	role: string;
	content: string;
	tool_calls?: unknown;
	tool_call_id?: string;
}

interface OpenAiToolCall {
	id: string;
	type: 'function';
	function: { name: string; arguments: string };
}

interface OpenAiResponse {
	id?: string;
	choices: Array<{
		message?: {
			role?: string;
			content?: string | null;
			tool_calls?: OpenAiToolCall[];
		};
		delta?: {
			role?: string;
			content?: string | null;
			tool_calls?: Array<{
				index: number;
				id?: string;
				function?: { name?: string; arguments?: string };
			}>;
		};
		finish_reason?: string | null;
	}>;
	usage?: {
		prompt_tokens?: number;
		completion_tokens?: number;
	};
}

function toOpenAiMessages(messages: Message[]): OpenAiMessage[] {
	const out: OpenAiMessage[] = [];
	for (const m of messages) {
		if (m.role === 'user' && 'toolCallId' in m && (m as { toolCallId?: string }).toolCallId) {
			const t = m as Message & { toolCallId: string; toolName: string };
			out.push({
				role: 'tool',
				tool_call_id: t.toolCallId,
				content: t.content,
			});
		} else {
			out.push({ role: m.role, content: m.content });
		}
	}
	return out;
}

export class OpenAiCompatibleProvider implements LLMProvider {
	readonly id: string;
	readonly displayName: string;
	readonly baseUrl: string;
	readonly defaultModel: string;
	readonly capabilities: ModelCapabilities;
	readonly anonymous: boolean;

	private readonly apiKey: string | undefined;
	private readonly headers: Record<string, string>;
	private readonly extraBody: Record<string, unknown>;
	private readonly embeddingModel: string | undefined;
	private readonly timeoutMs: number;
	private readonly transformRequest?: (body: Record<string, unknown>) => Record<string, unknown>;

	constructor(opts: OpenAiCompatOptions) {
		this.id = opts.id;
		this.displayName = opts.displayName;
		this.baseUrl = opts.baseUrl.replace(/\/+$/, '');
		this.defaultModel = opts.defaultModel;

		this.apiKey = opts.apiKey ?? (opts.apiKeyEnv ? process.env[opts.apiKeyEnv] : undefined);
		this.headers = { ...opts.headers };
		this.extraBody = { ...opts.extraBody };
		this.embeddingModel = opts.embeddingModel;
		this.timeoutMs = opts.timeoutMs ?? 120_000;
		this.transformRequest = opts.transformRequest;
		this.anonymous = opts.anonymous ?? false;

		this.capabilities = {
			toolCalling: true,
			contextWindow: 128_000,
			// Most OpenAI-compatible endpoints (OpenAI, OpenRouter, Groq,
			// Ollama, LM Studio, vLLM) expose /embeddings; Ollama serves it
			// from the same base URL as chat.
			embeddings: true,
			...opts.capabilities,
		};
	}

	/** Embed text inputs via the /embeddings endpoint (semantic memory). */
	async embed(request: EmbeddingsRequest): Promise<number[][]> {
		const model = request.model ?? this.embeddingModel ?? this.defaultModel;
		const url = `${this.baseUrl}/embeddings`;
		const headers: Record<string, string> = {
			'Content-Type': 'application/json',
			...this.headers,
		};
		if (this.apiKey) {
			headers['Authorization'] = bearerToken(this.apiKey);
		}

		const response = await httpJson<{
			data?: Array<{ index?: number; embedding?: number[] }>;
			error?: unknown;
		}>({
			url,
			method: 'POST',
			headers,
			body: { model, input: request.inputs },
			signal: request.signal,
			timeoutMs: this.timeoutMs,
		});

		if (!response?.data) {
			throw new ProviderError(
				`Embeddings request failed for ${this.id}: ${JSON.stringify(response?.error ?? 'no data')}`,
				0,
				this.id,
			);
		}
		// Preserve input order; some servers return embeddings out of order.
		const byIndex = new Map<number, number[]>();
		for (const item of response.data) {
			if (item.embedding) byIndex.set(item.index ?? byIndex.size, item.embedding);
		}
		return request.inputs.map((_, i) => byIndex.get(i) ?? []);
	}

	estimateCost(usage: TokenUsage): { input: number; output: number } {
		// Default rough pricing (USD per 1M tokens). Custom providers can
		// override via transformRequest or a subclass.
		return { input: 1.0, output: 3.0 };
	}

	async chat(request: ChatRequest): Promise<ChatResponse> {
		const body = this.buildBody(request);
		const url = `${this.baseUrl}/chat/completions`;
		const headers: Record<string, string> = {
			'Content-Type': 'application/json',
			...this.headers,
		};
		if (this.apiKey) {
			headers['Authorization'] = bearerToken(this.apiKey);
		}

		const response = await httpJson<OpenAiResponse>({
			url,
			method: 'POST',
			headers,
			body,
			signal: request.signal,
			timeoutMs: this.timeoutMs,
		});

		return this.parseResponse(response, request);
	}

	private buildBody(request: ChatRequest): Record<string, unknown> {
		const body: Record<string, unknown> = {
			model: request.model ?? this.defaultModel,
			messages: toOpenAiMessages(request.messages),
			stream: false,
		};
		if (request.temperature !== undefined) body['temperature'] = request.temperature;
		if (request.maxTokens !== undefined) body['max_tokens'] = request.maxTokens;
		if (request.stop !== undefined) body['stop'] = request.stop;
		if (request.tools && request.tools.length > 0) {
			body['tools'] = request.tools.map((t) => ({
				type: 'function',
				function: {
					name: t.name,
					description: t.description,
					parameters: t.inputSchema,
				},
			}));
			body['tool_choice'] = 'auto';
		}
		Object.assign(body, this.extraBody);
		return this.transformRequest ? this.transformRequest(body) : body;
	}

	private parseResponse(response: OpenAiResponse, request: ChatRequest): ChatResponse {
		const choice = response.choices?.[0];
		const content = choice?.message?.content ?? '';
		const toolCalls: ToolCall[] | undefined = choice?.message?.tool_calls?.map((tc) => ({
			id: tc.id,
			name: tc.function.name,
			arguments: tc.function.arguments,
		}));

		return {
			content,
			toolCalls: toolCalls && toolCalls.length > 0 ? toolCalls : undefined,
			usage: response.usage
				? {
					inputTokens: response.usage.prompt_tokens ?? 0,
					outputTokens: response.usage.completion_tokens ?? 0,
				}
				: undefined,
			model: request.model ?? this.defaultModel,
		};
	}

	/**
	 * Streaming chat via Server-Sent Events. Yields `text` deltas as tokens
	 * arrive and `toolCall` deltas when tool-call argument chunks are
	 * received (arguments are accumulated across chunks per tool index).
	 * Emits a final `usage` delta when the server reports usage in the last
	 * chunk (request `stream_options.include_usage` to opt in).
	 */
	async *stream(request: ChatRequest): AsyncGenerator<StreamDelta> {
		// Ask the server to include usage in the final chunk when supported.
		// Some OpenAI-compatible servers (older Ollama / LM Studio / custom
		// endpoints) reject unknown `stream_options` with a 400 — retry once
		// without it so streaming still works there.
		const buildBody = (includeUsage: boolean): Record<string, unknown> => {
			const body = this.buildBody(request);
			body['stream'] = true;
			if (includeUsage) body['stream_options'] = { include_usage: true };
			return body;
		};

		const headers: Record<string, string> = {
			'Content-Type': 'application/json',
			...this.headers,
		};
		if (this.apiKey) {
			headers['Authorization'] = bearerToken(this.apiKey);
		}

		const controller = new AbortController();
		const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
		const onAbort = () => controller.abort();
		request.signal?.addEventListener('abort', onAbort);

		// Accumulate tool-call argument chunks per tool index.
		const toolAcc = new Map<number, { id: string; name: string; arguments: string }>();

		try {
			let res = await this.streamFetch(buildBody(true), headers, controller.signal);
			// Strict servers reject stream_options: retry without it.
			if (res.status === 400) {
				res = await this.streamFetch(buildBody(false), headers, controller.signal);
			}
			if (!res.ok || !res.body) {
				const raw = await res.text().catch(() => '');
				throw new ProviderError(
					`HTTP ${res.status} from ${this.baseUrl}: ${truncate(raw, 500)}`,
					res.status,
					this.id,
					raw,
				);
			}

			const reader = res.body.getReader();
			const decoder = new TextDecoder();
			let buffer = '';
			while (true) {
				const { done, value } = await reader.read();
				if (done) break;
				buffer += decoder.decode(value, { stream: true });
				const lines = buffer.split('\n');
				buffer = lines.pop() ?? '';
				for (const line of lines) {
					const trimmed = line.trim();
					if (!trimmed.startsWith('data:')) continue;
					const data = trimmed.slice(5).trim();
					if (data === '[DONE]') return;
					try {
						const json = JSON.parse(data) as OpenAiResponse;
						if (json.usage) {
							yield {
								type: 'usage',
								usage: {
									inputTokens: json.usage.prompt_tokens ?? 0,
									outputTokens: json.usage.completion_tokens ?? 0,
								},
							};
						}
						const delta = json.choices?.[0]?.delta;
						if (delta?.content) {
							yield { type: 'text', text: delta.content };
						}
						if (delta?.tool_calls) {
							for (const tc of delta.tool_calls) {
								const acc = toolAcc.get(tc.index) ?? { id: '', name: '', arguments: '' };
								if (tc.id) acc.id = tc.id;
								if (tc.function?.name) acc.name += tc.function.name;
								if (tc.function?.arguments) acc.arguments += tc.function.arguments;
								toolAcc.set(tc.index, acc);
								yield {
									type: 'toolCall',
									toolCall: {
										id: acc.id || `call-${tc.index}`,
										name: acc.name || 'function',
										arguments: acc.arguments,
									},
								};
							}
						}
					} catch {
						// ignore partial lines
					}
				}
			}
		} finally {
			clearTimeout(timeout);
			request.signal?.removeEventListener('abort', onAbort);
		}
	}

	private async streamFetch(
		body: Record<string, unknown>,
		headers: Record<string, string>,
		signal: AbortSignal,
	): Promise<Response> {
		return fetch(`${this.baseUrl}/chat/completions`, {
			method: 'POST',
			headers,
			body: JSON.stringify(body),
			signal,
		});
	}
}

File: core/src/providers/openai_compat.rs
// Idexal Core — OpenAI-compatible provider
//
// One client serving OpenAI, OpenRouter, Groq, DeepSeek, Mistral, Together,
// Ollama, LM Studio, vLLM and any private gateway — they all speak the same
// /chat/completions wire format, so the only difference is base URL, key and
// headers (all user-configurable via config.rs, no vendor allowlist).
//
// Streaming detail carried over from the reference implementation: tool-call
// arguments arrive as fragments keyed by `index` and must be concatenated
// across chunks before the call is complete.

use super::types::{Delta, Message, Role, TokenUsage, ToolCall, ToolDefinition, Turn};
use crate::config::ProviderConfig;
use futures_util::StreamExt;
use serde::Deserialize;
use serde_json::json;

fn to_openai_messages(messages: &[Message]) -> Vec<serde_json::Value> {
    messages
        .iter()
        .map(|m| {
            if let Some(id) = &m.tool_call_id {
                return json!({ "role": "tool", "tool_call_id": id, "content": m.content });
            }
            match m.role {
                Role::System => json!({ "role": "system", "content": m.content }),
                Role::User => json!({ "role": "user", "content": m.content }),
                Role::Assistant => {
                    if m.tool_calls.is_empty() {
                        json!({ "role": "assistant", "content": m.content })
                    } else {
                        let calls: Vec<serde_json::Value> = m
                            .tool_calls
                            .iter()
                            .map(|tc| {
                                json!({
                                    "id": tc.id,
                                    "type": "function",
                                    "function": { "name": tc.name, "arguments": tc.arguments },
                                })
                            })
                            .collect();
                        json!({ "role": "assistant", "content": m.content, "tool_calls": calls })
                    }
                }
            }
        })
        .collect()
}

fn to_openai_tools(tools: &[ToolDefinition]) -> Vec<serde_json::Value> {
    tools
        .iter()
        .map(|t| {
            json!({
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.input_schema,
                }
            })
        })
        .collect()
}

#[derive(Deserialize, Debug, Default)]
struct FnDelta {
    name: Option<String>,
    arguments: Option<String>,
}

#[derive(Deserialize, Debug, Default)]
struct ToolCallDelta {
    index: usize,
    id: Option<String>,
    function: Option<FnDelta>,
}

#[derive(Deserialize, Debug, Default)]
struct MsgDelta {
    content: Option<String>,
    tool_calls: Option<Vec<ToolCallDelta>>,
}

#[derive(Deserialize, Debug, Default)]
struct Choice {
    delta: Option<MsgDelta>,
}

/// Sent once, in a final chunk whose `choices` array is empty — which is
/// why usage has to be read before the code gives up on a chunk for having
/// no choices.
#[derive(Deserialize, Debug, Default)]
struct UsageBlock {
    #[serde(default)]
    prompt_tokens: u64,
    #[serde(default)]
    completion_tokens: u64,
}

#[derive(Deserialize, Debug, Default)]
struct Chunk {
    choices: Option<Vec<Choice>>,
    usage: Option<UsageBlock>,
}

/// Accumulated state for one stream. Extracted from the network loop so the
/// chunk-ordering rules can be tested without a live endpoint.
#[derive(Default)]
struct StreamState {
    turn: Turn,
    /// index -> (id, name, accumulated arguments)
    pending: std::collections::HashMap<usize, (String, String, String)>,
    usage: Option<TokenUsage>,
}

impl StreamState {
    fn apply<F: FnMut(Delta)>(&mut self, chunk: Chunk, on_delta: &mut F) {
        if let Some(u) = chunk.usage {
            self.usage =
                Some(TokenUsage { input_tokens: u.prompt_tokens, output_tokens: u.completion_tokens });
        }
        let Some(delta) = chunk.choices.and_then(|c| c.into_iter().next()).and_then(|c| c.delta)
        else {
            return;
        };
        if let Some(text) = delta.content {
            if !text.is_empty() {
                self.turn.text.push_str(&text);
                on_delta(Delta::Text(text));
            }
        }
        if let Some(calls) = delta.tool_calls {
            for tc in calls {
                let entry = self
                    .pending
                    .entry(tc.index)
                    .or_insert_with(|| (String::new(), String::new(), String::new()));
                if let Some(id) = tc.id {
                    entry.0 = id;
                }
                if let Some(f) = tc.function {
                    if let Some(name) = f.name {
                        entry.1.push_str(&name);
                    }
                    if let Some(args) = f.arguments {
                        entry.2.push_str(&args);
                    }
                }
            }
        }
    }

    fn finish(mut self) -> Turn {
        let mut indices: Vec<usize> = self.pending.keys().copied().collect();
        indices.sort();
        for idx in indices {
            let (id, name, args) = self.pending.remove(&idx).expect("index came from the map");
            self.turn.tool_calls.push(ToolCall {
                // Some gateways stream tool calls without ever sending an
                // id; a stable synthetic one keeps the follow-up
                // tool_result matchable.
                id: if id.is_empty() { format!("call_{idx}") } else { id },
                name,
                arguments: if args.is_empty() { "{}".to_string() } else { args },
            });
        }
        self.turn.usage = self.usage;
        self.turn
    }
}

/// Build the request body. `include_usage` is separate because the same
/// body has to be re-sent without it when a server rejects the field.
fn build_body(
    cfg: &ProviderConfig,
    messages: &[Message],
    tools: &[ToolDefinition],
    include_usage: bool,
) -> serde_json::Value {
    let mut body = json!({
        "model": cfg.model,
        "stream": true,
        "messages": to_openai_messages(messages),
    });
    if include_usage {
        // Without this, a streaming response carries no token counts at all.
        body["stream_options"] = json!({ "include_usage": true });
    }
    if !tools.is_empty() {
        body["tools"] = json!(to_openai_tools(tools));
        body["tool_choice"] = json!("auto");
    }
    // extra_body is merged last so a user-configured value always wins,
    // including over stream_options.
    if let Some(obj) = body.as_object_mut() {
        for (k, v) in &cfg.extra_body {
            obj.insert(k.clone(), v.clone());
        }
    }
    body
}

async fn post(cfg: &ProviderConfig, body: &serde_json::Value) -> Result<reqwest::Response, String> {
    let client = reqwest::Client::new();
    let mut req = client
        .post(format!("{}/chat/completions", cfg.effective_base_url().trim_end_matches('/')))
        .header("content-type", "application/json");
    if let Some(key) = cfg.resolve_key() {
        req = req.header("authorization", format!("Bearer {key}"));
    }
    for (k, v) in &cfg.headers {
        req = req.header(k.as_str(), v.as_str());
    }
    req.json(body).send().await.map_err(|e| format!("request failed: {e}"))
}

pub async fn stream_turn<F: FnMut(Delta)>(
    cfg: &ProviderConfig,
    messages: &[Message],
    tools: &[ToolDefinition],
    mut on_delta: F,
) -> Result<Turn, String> {
    let mut res = post(cfg, &build_body(cfg, messages, tools, true)).await?;

    // Older Ollama and LM Studio builds reject the unknown `stream_options`
    // field with a 400. Token counts are worth having but never worth
    // failing a turn over, so drop the field and try once more — exactly
    // once, so a server that 400s for a real reason still surfaces it.
    if res.status() == reqwest::StatusCode::BAD_REQUEST {
        res = post(cfg, &build_body(cfg, messages, tools, false)).await?;
    }

    let status = res.status();
    if !status.is_success() {
        let text = res.text().await.unwrap_or_default();
        return Err(format!("HTTP {status}: {}", super::truncate(&text, 400)));
    }

    // Tool-call arguments arrive as fragments keyed by `index` and must be
    // concatenated across chunks before the call is complete.
    let mut state = StreamState::default();

    let mut stream = res.bytes_stream();
    let mut buffer = String::new();
    'outer: while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("stream read error: {e}"))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));
        while let Some(pos) = buffer.find('\n') {
            let line = buffer[..pos].trim_end_matches('\r').to_string();
            buffer.drain(..=pos);
            let Some(data) = line.strip_prefix("data:") else { continue };
            let data = data.trim();
            if data.is_empty() {
                continue;
            }
            if data == "[DONE]" {
                // Labelled: a plain break only leaves the line loop and
                // would leave us waiting on a socket with nothing to say.
                break 'outer;
            }
            let Ok(parsed) = serde_json::from_str::<Chunk>(data) else { continue };
            state.apply(parsed, &mut on_delta);
        }
    }

    Ok(state.finish())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::ProviderKind;

    fn consume(chunks: &[&str]) -> Turn {
        let mut state = StreamState::default();
        let mut sink = |_: Delta| {};
        for raw in chunks {
            let chunk = serde_json::from_str::<Chunk>(raw).expect("test fixture must parse");
            state.apply(chunk, &mut sink);
        }
        state.finish()
    }

    fn test_cfg() -> ProviderConfig {
        ProviderConfig {
            id: "t".into(),
            kind: ProviderKind::OpenaiCompatible,
            base_url: Some("http://localhost:11434/v1".into()),
            api_key: None,
            api_key_env: None,
            model: "m".into(),
            priority: 1,
            headers: Default::default(),
            extra_body: Default::default(),
            enabled: true,
        }
    }

    #[test]
    fn usage_is_requested_and_can_be_dropped_for_servers_that_reject_it() {
        // Some Ollama / LM Studio builds 400 on the unknown field, so the
        // retry body must be byte-for-byte the original minus that field.
        let cfg = test_cfg();
        let with = build_body(&cfg, &[Message::user("hi")], &[], true);
        assert_eq!(with["stream_options"]["include_usage"], true);

        let without = build_body(&cfg, &[Message::user("hi")], &[], false);
        assert!(without.get("stream_options").is_none());
        assert_eq!(without["messages"], with["messages"], "only the usage field differs");
    }

    #[test]
    fn usage_arrives_in_a_final_chunk_that_has_no_choices() {
        // Regression: reading `choices[0].delta` first and bailing when it
        // is missing skips the usage chunk entirely — the field would be
        // requested from the server and then silently thrown away.
        let turn = consume(&[
            r#"{"choices":[{"delta":{"content":"hello"}}]}"#,
            r#"{"choices":[],"usage":{"prompt_tokens":812,"completion_tokens":97,"total_tokens":909}}"#,
        ]);
        assert_eq!(turn.text, "hello");
        let usage = turn.usage.expect("usage must survive the empty-choices chunk");
        assert_eq!(usage.input_tokens, 812);
        assert_eq!(usage.output_tokens, 97);
    }

    #[test]
    fn a_server_that_never_reports_usage_yields_none() {
        let turn = consume(&[r#"{"choices":[{"delta":{"content":"hi"}}]}"#]);
        assert!(turn.usage.is_none(), "unreported must not look like zero");
    }

    #[test]
    fn tool_arguments_are_concatenated_across_chunks() {
        let turn = consume(&[
            r#"{"choices":[{"delta":{"tool_calls":[{"index":0,"id":"c1","function":{"name":"read_file","arguments":"{\"pa"}}]}}]}"#,
            r#"{"choices":[{"delta":{"tool_calls":[{"index":0,"function":{"arguments":"th\":\"a.txt\"}"}}]}}]}"#,
        ]);
        assert_eq!(turn.tool_calls.len(), 1);
        assert_eq!(turn.tool_calls[0].id, "c1");
        assert_eq!(turn.tool_calls[0].name, "read_file");
        assert_eq!(turn.tool_calls[0].arguments, r#"{"path":"a.txt"}"#);
    }

    #[test]
    fn tool_result_uses_the_tool_role() {
        let msgs = vec![Message::tool_result("call_1", "42")];
        let arr = to_openai_messages(&msgs);
        assert_eq!(arr[0]["role"], "tool");
        assert_eq!(arr[0]["tool_call_id"], "call_1");
        assert_eq!(arr[0]["content"], "42");
    }

    #[test]
    fn assistant_tool_calls_use_the_function_envelope() {
        let msgs = vec![Message::assistant(
            "",
            vec![ToolCall { id: "c1".into(), name: "list_dir".into(), arguments: r#"{"path":"."}"#.into() }],
        )];
        let arr = to_openai_messages(&msgs);
        assert_eq!(arr[0]["tool_calls"][0]["type"], "function");
        assert_eq!(arr[0]["tool_calls"][0]["function"]["name"], "list_dir");
    }

    #[test]
    fn system_message_stays_in_the_array() {
        let msgs = vec![Message {
            role: Role::System,
            content: "be brief".into(),
            tool_calls: vec![],
            tool_call_id: None,
        }];
        let arr = to_openai_messages(&msgs);
        assert_eq!(arr.len(), 1);
        assert_eq!(arr[0]["role"], "system");
    }
}
